TTS voices : persist default voice per language / refactor GUI

panaC opened the pull request:
Fixes #2811

panaC committed:
first commit

migrate ttsVoice to ttsVoices array

panaC committed:
fix: tts defaultVoices

try to apply throttle/debounce to ttsState with no success

panaC committed:
up

danielweck committed:
Merge branch 'develop' into fix/ttsvoices

danielweck committed:
Merge branch 'develop' into fix/ttsvoices

danielweck commented:
I reviewed the PR code...I am wondering about BCP47 language code comparisons, shouldn't the logic make sure to prioritise accurate complete matches, then fallback to prefix only?

For example, assuming `selectedLanguage.code` can be a full `en-US` or `en-UK` code, or a partial `en` code:

```js
const selectedVoice = defaultVoices.find(({language}) => language.split("-")[0].toLowerCase() === selectedLanguage.code);
```

===>

```js
const selectedVoice = defaultVoices.find(({language}) => language.toLowerCase() === selectedLanguage.code) || defaultVoices.find(({language}) => language.split("-")[0].toLowerCase() === selectedLanguage.code);
```

I ran a search for `split("-")` to find all ocurrences of language code comparisons, but I don't fully understand the possible values of the operands so I am not sure if the "prioritize-full + fallback-to-partial" logic is necessary.

danielweck commented:
There's a GUI refresh bug (multiple TTS pause + play sequences, ultra-repeated?) when selecting language and/or voice.
This doesn't occur when selecting speech rate / playback speed, or highlights style / colours, etc.

danielweck reviewed (changes requested):
GUI refresh bug

danielweck commented:
Other than the GUI refresh bug, I think this works. I tested The Waste Land on MacOS with different English and German voices (Greek doesn't work because they use a different language code, and on my computer I only have a single voice for Italian)

danielweck committed:
Merge branch 'develop' into fix/ttsvoices

danielweck committed:
fix build

panaC commented:
> I reviewed the PR code...I am wondering about BCP47 language code comparisons, shouldn't the logic make sure to prioritise accurate complete matches, then fallback to prefix only?
> 
> For example, assuming `selectedLanguage.code` can be a full `en-US` or `en-UK` code, or a partial `en` code:
> 
> ```js
> const selectedVoice = defaultVoices.find(({language}) => language.split("-")[0].toLowerCase() === selectedLanguage.code);
> ```
> 
> ===>
> 
> ```js
> const selectedVoice =
>        defaultVoices.find(({language}) => language.toLowerCase() === selectedLanguage.code)
>     || defaultVoices.find(({language}) => language.split("-")[0].toLowerCase() === selectedLanguage.code);
> ```
> 
> I ran a search for `split("-")` to find all ocurrences of language code comparisons, but I don't fully understand the possible values of the operands so I am not sure if the "prioritize-full + fallback-to-partial" logic is necessary.


we can do a full comparaison by security, but

`selectedLanguage.code` can only be of type lang: 'en','fr' (2 or 3 first character before the hyphen and lowercase) or of type 'US','FR' (2 last character after the hyphen and uppercase) 

cf: https://github.com/readium/speech/blob/0f17fc25e2956de5e6b1174fc9813c36fd8ff623/src/voices.ts#L249

danielweck commented:
try changing the TTS voice while playing ... the GUI flickers massively.

this doesn't happen when changing the speech rate, the highlights etc

panaC commented:
> try changing the TTS voice while playing ... the GUI flickers massively.
> 
> this doesn't happen when changing the speech rate, the highlights etc

For me (macbook m1), there is the same (huge!?) GUI flickering no matter what I change like speech rate or voice. I will investigate

danielweck commented:
I am seeing a duplicate Spanish voice on Mac, could you please test at your end?

<img width="680" alt="Screenshot 2025-03-03 at 16 49 14" src="https://github.com/user-attachments/assets/42a1ba65-a022-4766-af71-565e619c3634" />

panaC committed:
fix: ttsVoices combobox selectedKey (controlled) and fix doublon in readium/speech repo and move the popover tts settings above the TTS control node, VoiceSelection is not re-rendererd on ttsState quick mutation

panaC committed:
clean

panaC commented:
I'm merging now

panaC commented:
> The problem with the voice selector is that this often kills the TTS (stops playing, impossible to resume)

It seems to me that the issue is still present, let's continue to track the issue on a dedicated issue

danielweck commented:
Duplicate TTS voice fixed in readium-speech, thank you :)

<img width="946" alt="Screenshot 2025-03-04 at 10 07 35" src="https://github.com/user-attachments/assets/2fd82494-d230-4e10-b5f8-92f0c3232b13" />

danielweck commented:
I posted a new issue for the TTS configuration panel flickering: https://github.com/edrlab/thorium-reader/issues/2839
